vmalinovskiy: added code hints with examples generation

vmalinovskiy opened the pull request.

vmalinovskiy committed:
vmalinovskiy: added code hints with examples generation

TheFrozenFire commented on codeHintsInstall.php:
```diff
@@ -0,0 +1,295 @@
+$classDefinition = substr($classCode, strpos($classCode, "class Client extends SwaggerClient"));
+$classCode = $packageInfo."\n\n".$docText."\n".$classDefinition;
+file_put_contents($file, $classCode);
+echo file_get_contents($file);
```
Do we want this echoed on install?

TheFrozenFire reviewed (commented):
One concern I have with this is that it looks like it modifies the API client class file directly, instead of an unversioned file. This could result in failed updates if Composer is checking for unstaged changes in dependencies.

It would be preferable to have the API client implement an interface which we then supplement with the comments. We would then add this file to .gitignore.

TheFrozenFire commented on codeHintsInstall.php:
```diff
@@ -0,0 +1,295 @@
+}
+
+//UPDATING Client.php class to include documentation
+$file = './src/CallFire/Api/Client.php';
```
This path will resolve differently when the client is installed as a dependency. You might want to use \_\_DIR\_\_ as a reference point instead.

TheFrozenFire commented on codeHintsInstall.php:
```diff
@@ -0,0 +1,295 @@
+$packageInfo = substr($classCode, 0, strpos($classCode, $lastLineForPackagesDefinitions) + strlen($lastLineForPackagesDefinitions));
+$classDefinition = substr($classCode, strpos($classCode, "class Client extends SwaggerClient"));
+$classCode = $packageInfo."\n\n".$docText."\n".$classDefinition;
+file_put_contents($file, $classCode);
```
This will not be idempotent, I think. Each time the script is run, it seems like it will add more comments. Using a separate interface that you can replace in its entirety will make this much cleaner.

vmalinovskiy commented:
Yes, agree with you, will fix this and pull new version with interface.
Also will delete that last echo. It is not required for us.

On Tue, Feb 28, 2017 at 6:59 PM, Justin Martin <notifications@github.com>
wrote:

> *@TheFrozenFire* commented on this pull request.
>
> One concern I have with this is that it looks like it modifies the API
> client class file directly, instead of an unversioned file. This could
> result in failed updates if Composer is checking for unstaged changes in
> dependencies.
>
> It would be preferable to have the API client implement an interface which
> we then supplement with the comments. We would then add this file to
> .gitignore.
> ------------------------------
>
> In codeHintsInstall.php
> <https://github.com/CallFire/CallFire-PHP-SDK/pull/27#discussion_r103497627>
> :
>
> > +$paths = $decodedSpec['paths'];
> +foreach ($paths as $apiBunch) {
> +    foreach ($apiBunch as $apiInstance) {
> +        $docText .= "\n\n".loadDataToFile($apiInstance);
> +    }
> +}
> +
> +//UPDATING Client.php class to include documentation
> +$file = './src/CallFire/Api/Client.php';
> +$classCode = file_get_contents($file);
> +$lastLineForPackagesDefinitions = 'use Swagger\Client as SwaggerClient;';
> +$packageInfo = substr($classCode, 0, strpos($classCode, $lastLineForPackagesDefinitions) + strlen($lastLineForPackagesDefinitions));
> +$classDefinition = substr($classCode, strpos($classCode, "class Client extends SwaggerClient"));
> +$classCode = $packageInfo."\n\n".$docText."\n".$classDefinition;
> +file_put_contents($file, $classCode);
> +echo file_get_contents($file);
>
> Do we want this echoed on install?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/CallFire/CallFire-PHP-SDK/pull/27#pullrequestreview-24285855>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ANqruNGz-bMUxNO8sonKQTAzDS-NPMyqks5rhFJagaJpZM4MLOR3>
> .
>



-- 
*Vladimir Malinovskiy* Software Engineer
vmalinovskiy@callfire.com | 380632677617
email <vmalinovskiy@callfire.com> | website <https://www.callfire.com/> |
Facebook <https://www.facebook.com/callfire> | Twitter
<https://www.twitter.com/callfire>

vmalinovskiy commented:
Regarding this comment:
"path will resolve differently when the client is installed as a dependency."
- maybe will try to use fullPath to get access to Client.php, you are right.


"This will not be idempotent, I think. Each time the script is run, it
seems like it will add more comments." - nope, it will work correctly. I've
tested that and can approve that this regenerate existing docs, but in any
case this will be changed when interface will be added.
Will do changes as soon as possible.

On Tue, Feb 28, 2017 at 7:02 PM, Vladimir Malinovskiy <
vmalinovskiy@callfire.com> wrote:

> Yes, agree with you, will fix this and pull new version with interface.
> Also will delete that last echo. It is not required for us.
>
> On Tue, Feb 28, 2017 at 6:59 PM, Justin Martin <notifications@github.com>
> wrote:
>
>> *@TheFrozenFire* commented on this pull request.
>>
>> One concern I have with this is that it looks like it modifies the API
>> client class file directly, instead of an unversioned file. This could
>> result in failed updates if Composer is checking for unstaged changes in
>> dependencies.
>>
>> It would be preferable to have the API client implement an interface
>> which we then supplement with the comments. We would then add this file to
>> .gitignore.
>> ------------------------------
>>
>> In codeHintsInstall.php
>> <https://github.com/CallFire/CallFire-PHP-SDK/pull/27#discussion_r103497627>
>> :
>>
>> > +$paths = $decodedSpec['paths'];
>> +foreach ($paths as $apiBunch) {
>> +    foreach ($apiBunch as $apiInstance) {
>> +        $docText .= "\n\n".loadDataToFile($apiInstance);
>> +    }
>> +}
>> +
>> +//UPDATING Client.php class to include documentation
>> +$file = './src/CallFire/Api/Client.php';
>> +$classCode = file_get_contents($file);
>> +$lastLineForPackagesDefinitions = 'use Swagger\Client as SwaggerClient;';
>> +$packageInfo = substr($classCode, 0, strpos($classCode, $lastLineForPackagesDefinitions) + strlen($lastLineForPackagesDefinitions));
>> +$classDefinition = substr($classCode, strpos($classCode, "class Client extends SwaggerClient"));
>> +$classCode = $packageInfo."\n\n".$docText."\n".$classDefinition;
>> +file_put_contents($file, $classCode);
>> +echo file_get_contents($file);
>>
>> Do we want this echoed on install?
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/CallFire/CallFire-PHP-SDK/pull/27#pullrequestreview-24285855>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ANqruNGz-bMUxNO8sonKQTAzDS-NPMyqks5rhFJagaJpZM4MLOR3>
>> .
>>
>
>
>
> --
> *Vladimir Malinovskiy* Software Engineer
> vmalinovskiy@callfire.com | 380632677617
> email <vmalinovskiy@callfire.com> | website <https://www.callfire.com/> |
> Facebook <https://www.facebook.com/callfire> | Twitter
> <https://www.twitter.com/callfire>
>



-- 
*Vladimir Malinovskiy* Software Engineer
vmalinovskiy@callfire.com | 380632677617
email <vmalinovskiy@callfire.com> | website <https://www.callfire.com/> |
Facebook <https://www.facebook.com/callfire> | Twitter
<https://www.twitter.com/callfire>

vmalinovskiy committed:
vmalinovskiy: added code hints with examples generation (refactored)

vmalinovskiy commented:
Please check versions where your comments fixed.

vmalinovskiy committed:
vmalinovskiy: added main request types integration tests and updated code hints generation script

TheFrozenFire commented on .gitignore:
```diff
@@ -1,5 +1,4 @@
 vendor/
 test.php
-phpunit.xml
```
phpunit.xml should be in the .gitignore, using phpunit.xml.dist as the committed filename instead. This allows phpunit.xml.dist to be copied to phpunit.xml for local changes without getting accidentally committed. PHPUnit will read phpunit.xml.dist if phpunit.xml doesn't exist.

TheFrozenFire commented on composer.json:
```diff
@@ -12,11 +12,20 @@
+    },
+    "scripts": {
+        "post-install-cmd": [
+            "php codeHintsInstall.php"
```
Could we encapsulate the codeHintsInstall.php functionality in a class instead, and use a class method reference? [See here](https://getcomposer.org/doc/articles/scripts.md#defining-scripts). This will prevent any sort of issues with the working path not resolving (e.g. if composer install is run outside of the project directory).

TheFrozenFire commented on composer.json:
```diff
@@ -12,11 +12,20 @@
+        ]
+    },
+    "scripts": {
+        "post-install-cmd": [
```
Are we sure we want this to run on install? In certain situations this might cause errors that would break the install process (e.g. with a locked-down corporate firewall preventing swagger.json access).

It might be preferable to simply have the developer run this command manually and maybe include some post-install output to instruct them of this option.

Having a parameter for the command script to point to a custom swagger.json path might also be helpful.

TheFrozenFire commented on phpunit.xml:
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
File should be named phpunit.xml.dist.

TheFrozenFire commented on src/CallFire/Api/DocumentedClient.php:
```diff
@@ -0,0 +1,1819 @@
+<?php
```
Do we want to commit a pre-generated version of this? It'll become outdated fairly quickly, or we'll have to update it regularly.

vmalinovskiy replied:
Did that. No pregenerated version and customer will run generation manually when he wants.

TheFrozenFire commented on phpunit.xml:
```diff
@@ -0,0 +1,12 @@
+        </testsuite>
+    </testsuites>
+    <php>
+        <env name="API_Username" value="login" />
```
We should leave these credentials empty by default, so the integration tests don't run.

TheFrozenFire reviewed (changes requested):
Great work overall. :+1:

vmalinovskiy committed:
vmalinovskiy: moved code generation script into class method, turned off auto generation for code hints

vmalinovskiy committed:
vmalinovskiy: updated readmy with info how to generate code hints

vmalinovskiy committed:
vmalinovskiy: moved code generation script into class method, turned off auto generation for code hints

vmalinovskiy commented:
Done. Removed auto generation and moved script to Class. Added info in
readmy how to generate docs.

https://github.com/CallFire/CallFire-PHP-SDK/compare/2.0...vmalinovskiy:2.0?expand=1

Please review again.

On Sat, Mar 11, 2017 at 1:34 AM, Justin Martin <notifications@github.com>
wrote:

> *@TheFrozenFire* requested changes on this pull request.
>
> Great work overall. 👍
> ------------------------------
>
> In .gitignore
> <https://github.com/CallFire/CallFire-PHP-SDK/pull/27#discussion_r105508849>
> :
>
> > @@ -1,5 +1,4 @@
>  vendor/
>  test.php
> -phpunit.xml
>
> phpunit.xml should be in the .gitignore, using phpunit.xml.dist as the
> committed filename instead. This allows phpunit.xml.dist to be copied to
> phpunit.xml for local changes without getting accidentally committed.
> PHPUnit will read phpunit.xml.dist if phpunit.xml doesn't exist.
> ------------------------------
>
> In composer.json
> <https://github.com/CallFire/CallFire-PHP-SDK/pull/27#discussion_r105509508>
> :
>
> >      },
>      "autoload": {
>          "psr-0": {
>              "CallFire": "src/"
> -        }
> +        },
> +        "classmap": [
> +            "test/"
> +        ]
> +    },
> +    "scripts": {
> +        "post-install-cmd": [
> +            "php codeHintsInstall.php"
>
> Could we encapsulate the codeHintsInstall.php functionality in a class
> instead, and use a class method reference? See here
> <https://getcomposer.org/doc/articles/scripts.md#defining-scripts>. This
> will prevent any sort of issues with the working path not resolving (e.g.
> if composer install is run outside of the project directory).
> ------------------------------
>
> In composer.json
> <https://github.com/CallFire/CallFire-PHP-SDK/pull/27#discussion_r105509884>
> :
>
> >      },
>      "autoload": {
>          "psr-0": {
>              "CallFire": "src/"
> -        }
> +        },
> +        "classmap": [
> +            "test/"
> +        ]
> +    },
> +    "scripts": {
> +        "post-install-cmd": [
>
> Are we sure we want this to run on install? In certain situations this
> might cause errors that would break the install process (e.g. with a
> locked-down corporate firewall preventing swagger.json access).
>
> It might be preferable to simply have the developer run this command
> manually and maybe include some post-install output to instruct them of
> this option.
>
> Having a parameter for the command script to point to a custom
> swagger.json path might also be helpful.
> ------------------------------
>
> In phpunit.xml
> <https://github.com/CallFire/CallFire-PHP-SDK/pull/27#discussion_r105509998>
> :
>
> > @@ -0,0 +1,12 @@
> +<?xml version="1.0" encoding="UTF-8"?>
>
> File should be named phpunit.xml.dist.
> ------------------------------
>
> In src/CallFire/Api/DocumentedClient.php
> <https://github.com/CallFire/CallFire-PHP-SDK/pull/27#discussion_r105510133>
> :
>
> > @@ -0,0 +1,1819 @@
> +<?php
>
> Do we want to commit a pre-generated version of this? It'll become
> outdated fairly quickly, or we'll have to update it regularly.
> ------------------------------
>
> In phpunit.xml
> <https://github.com/CallFire/CallFire-PHP-SDK/pull/27#discussion_r105510790>
> :
>
> > @@ -0,0 +1,12 @@
> +<?xml version="1.0" encoding="UTF-8"?>
> +<phpunit bootstrap="vendor/autoload.php">
> +    <testsuites>
> +        <testsuite name="CallFire-PHP-SDK test">
> +            <directory>test</directory>
> +        </testsuite>
> +    </testsuites>
> +    <php>
> +        <env name="API_Username" value="login" />
>
> We should leave these credentials empty by default, so the integration
> tests don't run.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/CallFire/CallFire-PHP-SDK/pull/27#pullrequestreview-26398553>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ANqruA-IYPHXW_iP1uKqNFW_WJEq2vMuks5rkd4FgaJpZM4MLOR3>
> .
>



-- 
*Vladimir Malinovskiy* Software Engineer
vmalinovskiy@callfire.com | 380632677617
email <vmalinovskiy@callfire.com> | website <https://www.callfire.com/> |
Facebook <https://www.facebook.com/callfire> | Twitter
<https://www.twitter.com/callfire>

vmalinovskiy committed:
vmalinovskiy: updated composer.json to be able to run phpunit test from jenkins

vmalinovskiy committed:
vmalinovskiy: updated tests and docs with response params checking

vmalinovskiy committed:
vmalinovskiy: updated travis.yml to run tests using composer

vmalinovskiy committed:
vmalinovskiy: updated travis.yml to run tests for 5.5 and 5.6 php only

vmalinovskiy committed:
vmalinovskiy: updated travis.yml to remove code coverage as not required (we have only few basic tests)

vmalinovskiy committed:
vmalinovskiy: updated abstract test to be able to use environment variables from outside

vmalinovskiy committed:
vmalinovskiy: removed test from travis config

TheFrozenFire commented on .travis.yml:
```diff
@@ -1,10 +1,11 @@
+install:
+  - composer clear-cache
+  - composer install --prefer-source
+notifications:
```
Do you want to continue to be notified of these builds?

vmalinovskiy commented:
It doesn't make sense as we removed testing from there. Will remove and
commit.

On Wed, Apr 5, 2017 at 6:58 PM, Justin Martin <notifications@github.com>
wrote:

> *@TheFrozenFire* commented on this pull request.
> ------------------------------
>
> In .travis.yml
> <https://github.com/CallFire/CallFire-PHP-SDK/pull/27#discussion_r109956568>
> :
>
> >    - 5.5
> -  - 5.4
> -  - 5.3
> -before_script:
> -   - composer install
> -script: "phpunit --verbose --coverage-clover ./build/logs/clover.xml"
> -after_script:
> -  - php vendor/bin/coveralls -v
> +before_install:
> +  - composer self-update
> +install:
> +  - composer clear-cache
> +  - composer install --prefer-source
> +notifications:
>
> Do you want to continue to be notified of these builds?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/CallFire/CallFire-PHP-SDK/pull/27#pullrequestreview-31093089>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ANqruJOEdcTlnMqwWa9NOFDmMJe-O6MWks5rs7ovgaJpZM4MLOR3>
> .
>



-- 
*Vladimir Malinovskiy* Software Engineer
vmalinovskiy@callfire.com | 380632677617
email <vmalinovskiy@callfire.com> | website <https://www.callfire.com/> |
Facebook <https://www.facebook.com/callfire> | Twitter
<https://www.twitter.com/callfire>

vmalinovskiy committed:
Removed notification command from travis.yml
